bundles/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/handler/SpeedTestHandler.java

```diff
@@ -65,90 +63,83 @@ public SpeedTestHandler(Thing thing) {
-        updateStatus(ThingStatus.ONLINE);
-    }
-
-    private void startAutomaticRefresh() {
-        final int initialDelay = 2;
-        refreshTask = scheduler.scheduleWithFixedDelay(this::launchSpeedTest, initialDelay,
-                configuration.refreshInterval, TimeUnit.MINUTES);
-        logger.info("Start automatic in {} minutes, refreshed every {} minutes", initialDelay,
+        configuration = getConfigAs(SpeedTestConfiguration.class);
+        logger.info("Speedtests starts in {} minutes, then refreshes every {} minutes", configuration.initialDelay,
                 configuration.refreshInterval);
+        refreshTask = scheduler.scheduleWithFixedDelay(this::startSpeedTest, configuration.initialDelay,
+                configuration.refreshInterval, TimeUnit.MINUTES);
+        updateStatus(ThingStatus.ONLINE);
     }
 
-    private void launchSpeedTest() {
-        updateState(CHANNEL_TEST_ISRUNNING, OnOffType.ON);
-        updateState(CHANNEL_TEST_START, new DateTimeType());
-        updateState(CHANNEL_TEST_END, UnDefType.NULL);
+    synchronized private void startSpeedTest() {
         if (speedTestSocket == null) {
-            speedTestSocket = new SpeedTestSocket(1500);
-            speedTestSocket.addSpeedTestListener(this);
-            startTest(SpeedTestMode.DOWNLOAD);
+            logger.debug("Network speedtest started");
+            final SpeedTestSocket socket = new SpeedTestSocket(1500);
+            speedTestSocket = socket;
+            socket.addSpeedTestListener(this);
+            updateState(CHANNEL_TEST_ISRUNNING, OnOffType.ON);
+            updateState(CHANNEL_TEST_START, new DateTimeType());
+            updateState(CHANNEL_TEST_END, UnDefType.NULL);
+            updateProgress(new QuantityType<>(0, SmartHomeUnits.PERCENT));
+            socket.startDownload(configuration.getDownloadURL());
         } else {
-            logger.warn("A speedtest is still in progress, will retry on next refresh");
+            logger.info("A speedtest is already in progress, will retry on next refresh");
         }
-
     }
 
-    private void startTest(final SpeedTestMode mode) {
-        updateProgress(new QuantityType<>(0, SmartHomeUnits.PERCENT));
-        String fullURL = configuration.url;
-        fullURL += fullURL.endsWith("/") ? "" : "/";
-        if (mode == SpeedTestMode.DOWNLOAD) {
-            fullURL += configuration.fileName;
-            speedTestSocket.startDownload(fullURL);
-        } else {
-            speedTestSocket.startUpload(fullURL, configuration.uploadSize);
+    synchronized private void stopSpeedTest() {
+        updateState(CHANNEL_TEST_ISRUNNING, OnOffType.OFF);
+        updateProgress(UnDefType.NULL);
+        updateState(CHANNEL_TEST_END, new DateTimeType());
+        if (speedTestSocket != null) {
+            SpeedTestSocket socket = speedTestSocket;
+            socket.closeSocket();
+            socket.removeSpeedTestListener(this);
+            socket = null;
```

Comment:
You need to set the `speedTestSocket` to null, too. Otherwise it will never be null.

Reply:
well spotted 👍